tests/sgx_zerobase/enc/enc.cpp

```diff
@@ -3,42 +3,87 @@
 
 #include <openenclave/enclave.h>
 #include <openenclave/internal/tests.h>
+#include <iostream>
+
 #include "sgx_zerobase_t.h"
 
-const char* protected_message = "Hello world from Enclave\n\0";
+static uint64_t faulting_address;
+static uint32_t error_code;
+static uint32_t exception_code;
 
-int secure_string_patching(
-    const char* source,
-    char* output,
-    size_t output_length)
+uint64_t test_pfgp_handler(oe_exception_record_t* exception_record)
 {
-    const char* running_source = source;
-    size_t running_length = output_length;
-    while (running_length > 0 && *running_source != '\0')
+    if (exception_record->code == OE_EXCEPTION_PAGE_FAULT)
+    {
+        const int MOV_INSTRUCTION_BYTES = 3;
+        faulting_address = exception_record->faulting_address;
+        error_code = exception_record->error_code;
+        exception_record->context->rip += MOV_INSTRUCTION_BYTES;
+        exception_code = OE_EXCEPTION_PAGE_FAULT;
+    }
+    else if (exception_record->code == OE_EXCEPTION_ACCESS_VIOLATION)
     {
-        *output = *running_source;
-        running_length--;
-        running_source++;
-        output++;
+        const int LGDT_INSTRUCTION_BYTES = 8;
+        faulting_address = exception_record->faulting_address;
+        error_code = exception_record->error_code;
+        exception_record->context->rip += LGDT_INSTRUCTION_BYTES;
+        exception_code = OE_EXCEPTION_ACCESS_VIOLATION;
     }
-    const char* ptr = protected_message;
-    while (running_length > 0 && *ptr != '\0')
+    else
     {
-        *output = *ptr;
-        running_length--;
-        ptr++;
-        output++;
+        /* Unexpected code */
+        oe_abort();
     }
-    if (running_length < 1)
+
+    return OE_EXCEPTION_CONTINUE_EXECUTION;
+}
+
+int test_enclave_memory_access(uint64_t address, struct exceptions* exception)
+{
+    oe_result_t result = OE_OK;
+
+    if (exception)
+        result = oe_add_vectored_exception_handler(false, test_pfgp_handler);
+
+    if (result != OE_OK)
     {
         return -1;
     }
-    *output = '\0';
-    int rval = -1;
-    OE_TEST(
-        unsecure_string_patching(&rval, source, output, output_length) ==
-        OE_OK);
-    return rval;
+
+    /* Assign value to memory address 'address'*/
+    asm volatile("mov %0, %%r8\n\t"
```

Comment:
Since `r8` is being used, you need to mark it as clobbered in the clobber list of the inline assembly.
Alternatively the following might work, since %0 is a register.
```c
asm volatile("mov %0, (%0)\n\t"
                     : "memory"
                     : "r"(address));
```